File: OctoberChallenge/week1/k-diff-pairs.py
class Solution(object):
    def findPairs(self, nums, k):
        """
        O(n) time implementation
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        count = 0;
        nums.sort();

        j = 0;
        tmp = nums[-1];
        for i in range(len(nums)):
            while (nums[i]-nums[j] > k) and j < i:
                j += 1;

            if (nums[i]-nums[j] == k) and tmp != nums[j] and j != i:
                count += 1;
                nums[j] = tmp;

        return count;

File: October2020Challenge/week1/k-diff-pairs.cpp
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        int count = 0;

        // Ensure numbers are small to large:
        sort(nums.begin(), nums.end());

        int j = 0;
        int tmp = INT_MAX;

        for(int i=0; i < nums.size(); i++) {
            // Increment-iterate j if condition doesn't meet
            while (j < i && (nums[i]-nums[j]>k)) j++;

            // Increase count iff non-self and diffK condition meets
            if(j != i && tmp != nums[j] && (nums[i]-nums[j]==k)) {
                count++;
                tmp = nums[j];
            }
        }

        return count;
    }
};
